### PR TITLE
Added multi WiFi saving feature

### PR DESCRIPTION
**Pull request by JPZV**
(Note: This is a rebase of #3703 with the 0_15 branch)

## TL;DR:

This PR adds the Multi-Wifi saving feature, which was requested and discussed in #2845, #2974, #852, #1228, and other issues that I may not found.

### User's side
For the user I tried to make it as simple as possible: Now the WiFi setting page will have three "new" sections:
- _Connect to a new network_: Here the user can scan for a nearby network or add it manually with the option "_Other network..._"
- _Saved networks_: Here the user can manage its saved network by changing the SSID, the Password, or deleting it.
- _Addresses settings_: Here is the old DHCP/IP and mDNS settings. Sadly, this will affect to all networks, so, for example, setting a static IP to 192.168.0.64/24 while on _networkA,_ this will be used for every other network, even if the subnet of _networkB_ starts from 192.168.1.0/24. So, DHCP is fully recommended from now on if using multiple networks.

### In back-end
Nowadays WLED saves the WiFi setting in two global vars: _clientSSID_ and _clientPass_. Both are strings (well... Char array). Now, I removed these vars and created three new ones: _clientNetsSSID_, _clientNetsPass_ and _clientSavedNets_. The first two are almost the same as the old vars, but now they're Strings Array (or Char Matrix, or Char Array Array, or whatever you want to call it), which can save up to 5 Networks in ESP32, or 3 in ESP8266 (those limits were selected arbitrarily). While _clientSavedNets_ is a Byte that stores how many networks are saved and loaded into memory.

When WLED wants to connect to a network, it first scan for every nearby networks and start looking for the nearest saved network. Then, tries to connect to it, and if it can't, then try with the next near one. Each attempt has like 5 seconds before failing. This amount of time was selected because "Last reconnect too old" message was being fired immediately with anything greater than 5 seconds and with just one network.

### Backward compatibility
Now WLED will check if the config file has a network saved with the last scheme, and if there's, then it will add it as the _first network_ and then it'll save the config file with the new scheme

### Drawbacks
Because I'm not familiar with **Improv** and **EEPROM saving** in WLED, and because of backward compatibility concerns, in both cases it'll set/load the first network. For example:

```diff
-  readStringFromEEPROM(  0, clientSSID, 32);
-  readStringFromEEPROM( 32, clientPass, 64);
-  readStringFromEEPROM( 96,      cmDNS, 32);
+  readStringFromEEPROM(  0, clientNetsSSID[0], 32);
+  readStringFromEEPROM( 32, clientNetsPass[0], 64);
+  readStringFromEEPROM( 96,             cmDNS, 32);
```

This shouldn't be a big issue for most users, except for those who uses Improv/EEPROM and wanted to save multiple WiFi networks.

Also, as mentioned before, now the IP Settings will affect to every network, even if the subnets differs com each others.

## Files Changed:

- **cfg.cpp**: Now the WiFi settings will be saved as `ssid%d` and `psk%d`, where `%d` goes from 0 to 4, instead of `ssid` and `psk` respectively
- **const.h**: Added `WLED_MAX_SAVED_NETWORKS`, `SUBPAGE_MAIN_SETTINGS`, and `SUBPAGE_WIFI_NETWORKS` constant
- **data/settings_wifi.htm**: Added three new sections: _Connect to a new network_, _Saved networks_, and _Addresses settings_. Also, changed the `loadJS()` function to letting load more than one JS.
- **data/style.css**: Added `input[type="password"]` to the same `input[type="text"]` style.
- **fcn_declare.h**: Changed the declaration of `getSettingsJS` to `void getSettingsJS(byte subPage, char* dest, byte subSettings);`, so it can load sub-settings from the same page.
- **improv.cpp**, **set.cpp**, **wled.cpp** and **wled_eeprom.cpp**: Added support for multiple WiFi settings.
- **wled.h**:
  - Replaced `clientSSID` and `clientPass` global vars with `clientNetsSSID` and `clientNetsPass` respectively.
  - Added `clientSavedNets` global var.
  - Now `WLED_WIFI_CONFIGURED` will be true if the first network is configured, or if `clientSavedNets` is greater than zero.
- **wled_server.cpp**: Now the function `serveSettingsJS` will accept the `s` argument for loading sub-settings from the same page.
  - In case `s` is zero (default behavior). It'll return the `GetV` JS function. Otherwise, it'll return the `GetW` one.
- **xml.cpp**: Added support for `SUBPAGE_WIFI_NETWORKS` sub-setting in `SUBPAGE_WIFI` page. In this case, it'll return a tuple of three where the first item is the amount of saved networks, while the second item is the maximum amount of networks that can be saved, and the last item is an array of every saved network.

### CODE REVIEW

**Comment by blazoncek** [comment]
Thank you. Will have a look ASAP.

**Review by blazoncek — CHANGES_REQUESTED** [review]
The idea is nice but I find implementation a bit inefficient.
There are multiple areas that need addressing.

I particularly do not like static memory allocation and handling of JSON. Unfortunately I did not delve in JavaScript and settings page due to lack of time. I will comment on those at some later time.

File: wled00/wled_server.cpp [hunk 1]
```diff
@@ -520,8 +520,13 @@ void serveSettingsJS(AsyncWebServerRequest* request)
     request->send(401, "application/javascript", buf);
     return;
   }
-  strcat_P(buf,PSTR("function GetV(){var d=document;"));
-  getSettingsJS(subPage, buf+strlen(buf));  // this may overflow by 35bytes!!!
+  byte subSettings = request->arg(F("s")).toInt();
```
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/wled_server.cpp:523]: This is completely necessary if you don't want to break the others settings pages.

Because on the webpage the networks are created dynamically, you cannot set them with GetV, because it'll fail with "unknown element CS{number}", but because we don't know the number of saved network before calling GetV, we cannot create them beforehand.

File: wled00/cfg.cpp [hunk 2]
```diff
@@ -41,11 +41,32 @@ bool deserializeConfig(JsonObject doc, bool fromFS) {
 #endif
 
   JsonObject nw_ins_0 = nw["ins"][0];
-  getStringFromJson(clientSSID, nw_ins_0[F("ssid")], 33);
```
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/cfg.cpp:44]: Thanks, I didn't think about making it as the others settings. Implemented in [75a69f7ea30ad6aac844b8b32c053924868124e2](https://github.com/Aircoookie/WLED/pull/3705/commits/75a69f7ea30ad6aac844b8b32c053924868124e2).

Sadly, with that, it lost the backward compatibility with the addresses settings, so everyone will start with "0,0,0,0" after updating.

File: wled00/cfg.cpp [hunk 3]
```diff
@@ -667,8 +688,13 @@ void serializeConfig() {
   JsonArray nw_ins = nw.createNestedArray("ins");
 
   JsonObject nw_ins_0 = nw_ins.createNestedObject();
```
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/cfg.cpp]: Implemented in [75a69f7ea30ad6aac844b8b32c053924868124e2](https://github.com/Aircoookie/WLED/pull/3705/commits/75a69f7ea30ad6aac844b8b32c053924868124e2).

File: wled00/const.h [hunk 4]
```diff
@@ -106,6 +106,15 @@
   #endif
 #endif
 
+// This max was selected arbitrarily. Should be adjusted for real-world cases
+#ifndef WLED_MAX_SAVED_NETWORKS
```
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/const.h:113]: Implemented in [13a0f77a086ec37dae4828a250102112e308f443](https://github.com/Aircoookie/WLED/pull/3705/commits/13a0f77a086ec37dae4828a250102112e308f443).

File: wled00/wled.h [hunk 5]
```diff
@@ -304,8 +310,9 @@ WLED_GLOBAL int8_t irPin _INIT(IRPIN);
 WLED_GLOBAL char ntpServerName[33] _INIT("0.wled.pool.ntp.org");   // NTP server to use
 
 // WiFi CONFIG (all these can be changed via web UI, no need to set them here)
-WLED_GLOBAL char clientSSID[33] _INIT(CLIENT_SSID);
-WLED_GLOBAL char clientPass[65] _INIT(CLIENT_PASS);
+WLED_GLOBAL byte clientSavedNets _INIT(CLIENT_SAVED_NETS);
+WLED_GLOBAL char clientNetsSSID[WLED_MAX_SAVED_NETWORKS][33] _INIT_N(({CLIENT_SSID}));
```
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/wled.h]: I knew making static will take a lot of heap, but making it dynamically was so awful. Anyway, I implemented in [75a69f7ea30ad6aac844b8b32c053924868124e2](https://github.com/Aircoookie/WLED/pull/3705/commits/75a69f7ea30ad6aac844b8b32c053924868124e2).

File: wled00/cfg.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/cfg.cpp:44]: Backwards (in)compatibility shouldn't be the case as "ins" is an array of objects and each object stores complete WiFi configuration.
There should always be at least one object in the array.

File: wled00/wled_server.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (blazoncek) [reply wled00/wled_server.cpp:523]: The same principle should be used as in LED settings for adding Color Order Overrides or LED outputs.

You should dynamically generate the number of networks within JS.

File: wled00/cfg.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/cfg.cpp:44]: But then you would have more memory allocated for storing multiple IP Addresses, and making it even more complex for the user to configure them.

File: wled00/wled.cpp [hunk 6]
```diff
@@ -743,15 +743,44 @@ void WLED::initConnection()
...
+
+            DEBUG_PRINT(F("Connecting to "));
+            DEBUG_PRINT(clientNetsSSID[i]);
+            DEBUG_PRINT("... ");
+
+            WiFi.begin(clientNetsSSID[i], clientNetsPass[i]);
+            int delay_count = 0;
+            // I was thinking about 10 seconds, but "Last reconnect too old" message was being fired immediately
+            while (WiFi.status() == WL_IDLE_STATUS && delay_count++ < 50) {
+                delay(100);
```
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/wled.cpp]: Implemented in [2619bb502ab7093629a147ebbaf2d11c51800ed2](https://github.com/Aircoookie/WLED/pull/3705/commits/2619bb502ab7093629a147ebbaf2d11c51800ed2)

File: wled00/cfg.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/cfg.cpp:44]: Implemented in [d360ebc1952058a3346ec2ca2a9bb0b9644fbffe](https://github.com/Aircoookie/WLED/pull/3705/commits/d360ebc1952058a3346ec2ca2a9bb0b9644fbffe)

File: wled00/wled_server.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (JPZV) [reply wled00/wled_server.cpp]: Implemented in [4a83c0d8dd917af47d42be1838dfcf60b2a7ceb1](https://github.com/Aircoookie/WLED/pull/3705/commits/4a83c0d8dd917af47d42be1838dfcf60b2a7ceb1)

**Comment by JPZV** [comment]
@blazoncek I've implemented all the changes that you requested until now.

Feel free to add more feedback to it.

**Comment by blazoncek** [comment]
I am sorry if this looks sublime, but it took me less time to implement than to describe what I wanted.

This is how it should look in UI.
Screenshot 2024-01-20 at 22 58 09
I've implemented vectors in the background so the memory is allocated dynamically for each configuration.

I've also added custom DNS which was long overdue.

I'll push a new branch so that we can compare and only take the best of both versions.

**Comment by blazoncek** [comment]
Closed in favour of #3709
